### PR TITLE
Frontend: set up output file .swiftsourceinfo

### PR DESCRIPTION
**Pull request by nkcsgexi**
This patch will focus on teaching driver and frontend to emit this file.
The actual content and de-serialization parts will come later.

More details: https://forums.swift.org/t/proposal-emitting-source-information-file-during-compilation/28794

### CODE REVIEW

**Comment by nkcsgexi** [comment]
@swift-ci Please Build Toolchain macOS Platform

**Comment by nkcsgexi** [comment]
@swift-ci Please Build Toolchain Linux Platform

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci Please Build Toolchain macOS Platform

**Comment by nkcsgexi** [comment]
@swift-ci Please Build Toolchain macOS Platform

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

File: lib/Driver/Driver.cpp [hunk 1]
```diff
@@ -2786,6 +2789,41 @@ void Driver::chooseSwiftModuleOutputPath(Compilation &C,
   }
 }
 
+void Driver::choosePrivateOutputFilePath(Compilation &C,
```
Comment (jrose-apple) [inline lib/Driver/Driver.cpp]: Nitpick: can you put "module" in the name of this function somewhere? I don't know if all private outputs belong in the module folder.

File: lib/Driver/Driver.cpp [hunk 2]
```diff
@@ -2786,6 +2789,41 @@ void Driver::chooseSwiftModuleOutputPath(Compilation &C,
...
+    Output->setAdditionalOutputForType(fileID, OFMOutputPath);
+  } else if (Output->getPrimaryOutputType() != file_types::TY_Nothing) {
+    auto ModulePath = Output->getAnyOutputForType(file_types::TY_SwiftModuleFile);
+    bool isTempFile = C.isTemporaryFile(ModulePath);
+    auto ModuleName = llvm::sys::path::filename(ModulePath);
+    llvm::SmallString<128> Path(llvm::sys::path::parent_path(ModulePath));
+    llvm::sys::path::append(Path, "Private");
+    // If the build system has created a Private dir for us to include the file, use it.
+    if (!llvm::sys::fs::exists(Path)) {
+      llvm::sys::path::remove_filename(Path);
```
Comment (jrose-apple) [inline lib/Driver/Driver.cpp]: Ah, clever. And I guess we can look in both places?

File: lib/Driver/ToolChains.cpp [hunk 3]
```diff
@@ -907,6 +912,8 @@ ToolChain::constructInvocation(const MergeModuleJobAction &job,
 
   addOutputsOfType(Arguments, context.Output, context.Args,
                    file_types::TY_SwiftModuleDocFile, "-emit-module-doc-path");
+  addOutputsOfType(Arguments, context.Output, context.Args,
+                   file_types::TY_SwiftSourceInfoFile, "-emit-module-source-info-path");
```
Comment (jrose-apple) [inline lib/Driver/ToolChains.cpp]: Nitpick: 80 columns.

File: lib/Frontend/FrontendInputsAndOutputs.cpp [hunk 4]
```diff
@@ -435,6 +435,12 @@ bool FrontendInputsAndOutputs::hasModuleDocOutputPath() const {
         return outs.ModuleDocOutputPath;
       });
 }
+bool FrontendInputsAndOutputs::hasModuleSourceInfoOutputPath() const {
+  return hasSupplementaryOutputPath(
+  [](const SupplementaryOutputPaths &outs) -> const std::string & {
```
Comment (jrose-apple) [inline lib/Frontend/FrontendInputsAndOutputs.cpp]: Nitpick: funny indentation here.

File: lib/Serialization/Serialization.cpp [hunk 5]
```diff
@@ -4774,6 +4775,22 @@ void swift::serializeToBuffers(
       *moduleDocBuffer = llvm::make_unique<llvm::SmallVectorMemoryBuffer>(
                            std::move(buf), options.DocOutputPath);
   }
+
+  if (!StringRef::withNullAsEmpty(options.SourceInfoOutputPath).empty()) {
+    SharedTimer timer("Serialization, swiftsourceinfo, to buffer");
+    llvm::SmallString<1024> buf;
+    llvm::raw_svector_ostream stream(buf);
+    writeSourceInfoToStream(stream, DC);
```
Comment (jrose-apple) [inline lib/Serialization/Serialization.cpp]: @harlanhaskins This always using a buffer is unfortunate; ideally we'd only use it when we're going to keep it in memory.

File: lib/Serialization/SerializeDoc.cpp [hunk 6]
```diff
@@ -502,3 +502,7 @@ void serialization::writeDocToStream(raw_ostream &os, ModuleOrSourceFile DC,
 
   S.writeToStream(os);
 }
+
+void serialization::writeSourceInfoToStream(raw_ostream &os, ModuleOrSourceFile DC) {
+  os << "Some stuff";
```
Comment (jrose-apple) [inline lib/Serialization/SerializeDoc.cpp]: Excellent source info. :-)

File: test/Driver/sourceinfo_file.swift [hunk 7]
```diff
@@ -0,0 +1,11 @@
+// RUN: %empty-directory(%t)
+// RUN: mkdir -p %t/build/Private
+// RUN: %swiftc_driver -driver-print-jobs -emit-module %s -emit-module-path %t/build/sourceinfo_file.swiftmodule -module-name sourceinfo_file | %FileCheck %s -check-prefix CHECK-PRIVATE
+
+// CHECK-PRIVATE: Private/sourceinfo_file.swiftsourceinfo
```
Comment (jrose-apple) [inline test/Driver/sourceinfo_file.swift]: Can you check that this is `build/Private/sourceinfo_file.swiftsourceinfo`?

File: lib/Driver/Driver.cpp (hunk 1 above)
Comment (jrose-apple) [reply lib/Driver/Driver.cpp]: Alternately, is it possible to combine this with the swiftdoc helper and just have a flag to say "Private" or not? Duplicated code is code that gets out of sync.

**Review by jrose-apple — COMMENTED** [review]
Looks pretty good, also not happy how much it took. cc @davidungar for later thought on how to streamline some of this.

File: lib/Driver/Driver.cpp (hunk 2 above)
Comment (nkcsgexi) [reply lib/Driver/Driver.cpp]: yeah, the source info loader should look into both places and prefer the `Private` one.

File: lib/Driver/Driver.cpp (hunk 1 above)
Comment (nkcsgexi) [reply lib/Driver/Driver.cpp]: Sounds good! Will do.

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci Please Build Toolchain macOS Platform

**Comment by nkcsgexi** [comment]
@swift-ci Please Build Toolchain Linux Platform

**Comment by nkcsgexi** [comment]
@swift-ci Please Build Toolchain macOS Platform

**Review by jrose-apple — APPROVED** [review]

**Comment by nkcsgexi** [comment]
@swift-ci please test

**Comment by nkcsgexi** [comment]
@swift-ci please test

**Comment by nkcsgexi** [comment]
@swift-ci Please test OS X platform

**Comment by davidungar** [comment]
LGTM

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci Please smoke test

**Comment by compnerd** [comment]
@nkcsgexi I think that this may have regressed the windows build: https://dev.azure.com/compnerd/windows-swift/_build/results?buildId=9686&view=logs&j=2d2b3007-3c5c-5840-9bb0-2b1ea49925f3&t=2fdae087-0aee-589c-2ab0-2b8fc539478d&l=184

**Pull request merged**